### PR TITLE
Enemy progression system

### PR DESCRIPTION
**Pull request by ShovelD**

### CODE REVIEW

File: Tank Survivors/Assets/Scripts/Enemies/Drone.cs [hunk 1]
```diff
@@ -47,6 +48,7 @@ public class Drone : SerializedMonoBehaviour, IEnemy
 
         [OdinSerialize]
         public string EnemyName { get; private set; }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Drone.cs:50]: Do we really need `EnemyName` ?

File: Tank Survivors/Assets/Scripts/Enemies/EnemyProducers/ConstantEnemyProducer.cs [hunk 2]
```diff
@@ -1,56 +1,46 @@
...
+        public override float EndTime => float.PositiveInfinity;
 
-        public void Produce(TankImpl tank, Transform enemyRoot)
+        public override void Produce(TankImpl tank, Transform enemyRoot)
         {
             timer -= Time.deltaTime;
             if (timer < 0)
             {
-                UnityEngine
+                var enemy = UnityEngine
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/EnemyProducers/ConstantEnemyProducer.cs:34]: Chek formatting, var shouldn't be used

File: Tank Survivors/Assets/Scripts/Enemies/Modules.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using Common;
+using Sirenix.OdinInspector;
+using Sirenix.Serialization;
+
+namespace Enemies
+{
+    public interface IModule : ICloneable { }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Modules.cs:8]: Why you use `ICloneable`

File: Tank Survivors/Assets/Scripts/Enemies/Modules.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        [FoldoutGroup("Health")]
+        public ModifiableValue<float> Health { get; set; } = new();
+
+        public object Clone()
+        {
+            HealthModule module = new HealthModule();
+            module.Health.SourceValue = Health.SourceValue;
+            module.Health.Modifications.AddRange(Health.Modifications);
+            return module;
+        }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/Modules.cs:43]: using `object` very bad, find other way to do what you need

File: Tank Survivors/Assets/Scripts/General/EnemyProgressor.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                    foreach (
+                        IModuleUpgrade upgrade in producer.ProgressorProperties.UpgradebleModules
+                    )
+                    {
+                        upgrade.ApplyUpgrade(producer);
+                    }
+                    producer.ProgressorProperties.LastUpdateTime = timer.CurrentTime;
+                }
+            }
+        }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/General/EnemyProgressor.cs:49]: Why you update enemies in update? Just update them when you creating them.

**Review by Kiuh — CHANGES_REQUESTED** [review]
image

- Move producer name inside foldout group and if possible, assign it when you choose producer type.
- Move modules inside a prefab and don't show it in this place
- Make start and end radius as external link to SO and create small, normal and bug radius, and incapsulate get creating point into that configs

image

- Move this progressor into other config, that can modify any of modules that he mentioned.
- In progressor config, make sure i can customize progression for every module separely.
- Make interval more intuitive (maybe in seconds and show readonly field in minutes)

File: Tank Survivors/Assets/Scripts/Enemies/Drone.cs (hunk 1 above)
Comment (ShovelD) [reply Tank Survivors/Assets/Scripts/Enemies/Drone.cs:50]: i've never added this property to interface. it appeared after one of your's commit

Comment (Kiuh) [reply Tank Survivors/Assets/Scripts/Enemies/Drone.cs:50]: So, remove it

File: Tank Survivors/Assets/Scripts/General/EnemyProgressor.cs [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                )
+                {
+                    foreach (IModuleUpgrade upgrade in producer.Progressor.UpgradebleModules)
+                    {
+                        upgrade.ApplyUpgrade(producer);
+                    }
+                    producer.Progressor.LastUpdateTime = timer.CurrentTime;
+                }
+            }
+        }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/General/EnemyProgressor.cs:61]: Little wrong way.
Better make producers ask for current time, they will calculate stats for enemy that they want to create depended on time and progression config that they have.
As I see, you will delete `LastUpdateTime`

**Review by Kiuh — CHANGES_REQUESTED** [review]

File: Tank Survivors/Assets/Scripts/Enemies/ModuleUpgrades.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        [OdinSerialize]
+        [EnumToggleButtons]
+        public MathOperation Operation { get; private set; } = MathOperation.Plus;
+        public abstract void ApplyUpgrade(IEnemyProducer producer);
+
+        protected T GetModule<T>(IEnemyProducer producer)
+            where T : class, IModule
+        {
+            T module = producer.Modules.GetConcrete<T, IModule>();
+            return module;
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/ModuleUpgrades.cs:35]:
```suggestion
            return producer.Modules.GetConcrete<T, IModule>();
```

File: Tank Survivors/Assets/Scripts/General/EnemyGenerator.cs [hunk 8]
```diff
@@ -22,7 +23,7 @@ public class EnemyGenerator : MonoBehaviour
         [SerializeField]
         private TankImpl tank;
 
-        private List<IEnemyProducer> enemyProducers;
+        private List<IEnemyProducer> enemyProducers = new();
         private List<IEnemyProducer> toRemove = new();
         private List<BossProducer> bossProducers;
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/General/EnemyGenerator.cs:28]:
```suggestion
        private List<BossProducer> bossProducers = new();
```

File: Tank Survivors/Assets/Scripts/Enemies/EnemyProducers/DefaultEnemyProducer.cs [hunk 9]
```diff
@@ -1,74 +1,67 @@
...
 
         private Vector3 GetRandomPoint()
         {
+            CircleZone boundsRadius = radius.GetCircleZone(radiusPreset);
             Vector2 point = UnityEngine.Random.insideUnitCircle;
-            return (point * (endCircleRadius - startCircleRadius))
-                + (point.normalized * startCircleRadius);
+            return (point * (boundsRadius.Max - boundsRadius.Min))
+                + (point.normalized * boundsRadius.Min);
         }
```
Comment (Kiuh) [inline Tank Survivors/Assets/Scripts/Enemies/EnemyProducers/DefaultEnemyProducer.cs:65]: Move this method into Circle zone

**Review by Kiuh — APPROVED** [review]
And check that game is correctly builds with no errors and works in build

**Pull request merged**